src/coreclr/jit/optimizebools.cpp

```diff
@@ -1570,3 +1570,698 @@ PhaseStatus Compiler::optOptimizeBools()
+                                if (currentPattern < optRngPattern.optGetMinPattern())
+                                {
+                                    // Update the min pattern and the minOp to the CNS_INT tree with the min pattern
+                                    optRngPattern.optSetMinPattern(currentPattern);
+                                    optRngPattern.optSetMinOp(condition1->gtGetOp2());
+                                }
+                                else if (currentPattern > optRngPattern.optGetMaxPattern())
+                                {
+                                    // Update the max pattern
+                                    optRngPattern.optSetMaxPattern(currentPattern);
+                                }
+                                patternIndex++;
+
+                                // Update the code offset range
+                                optRngPattern.optSetBbCodeOffsEnd(currBb->bbCodeOffsEnd);
+
+                                foundPattern = true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Stop searching if patterns have been found in previous BBs, but the current BB does not have a pattern
+        if (foundPattern && patternIndex < (int)(currBb->bbNum - firstPatternBBNum + 1))
+        {
+            break;
+        }
+
+        prevBb = currBb;
+    }
+
+    if (foundPattern)
+    {
+        int patternCount = patternIndex;
+        optRngPattern.optSetPatternCount(patternCount);
+
+#ifdef DEBUG
+        if (this->verbose)
+        {
+            printf("Reserved patterns:\n");
+            for (int idx = 0; idx < patternCount; idx++)
+            {
+                printf("%d ", optRngPattern.optGetPattern(idx));
+            }
+            printf("\n\n");
+        }
+#endif
+        // Find range of pattern values
+        int minPattern = optRngPattern.optGetMinPattern();
+        int maxPattern = optRngPattern.optGetMaxPattern();
+#ifdef DEBUG
+        if (this->verbose)
+        {
+            printf("Min pattern value: %d\n", minPattern);
+            printf("Max pattern value: %d\n", maxPattern);
+        }
+#endif // DEBUG
+
+        int rangePattern = maxPattern - minPattern + 1;
+#ifdef DEBUG
+        if (this->verbose)
+        {
+            printf("Range of pattern values: %d\n", rangePattern);
+        }
+#endif // DEBUG
+
+        if (rangePattern > optRngPattern.m_sizePatterns)
+        {
+            printf("Range of pattern values is too big (> %d): %d\n", optRngPattern.m_sizePatterns, rangePattern);
+            return PhaseStatus::MODIFIED_NOTHING;
+        }
+        optRngPattern.optSetRangePattern(rangePattern);
+
+        // Find bitmap of pattern values
+        unsigned int btPatterns = 0;
+        for (int idxPattern = 0; idxPattern < patternCount; idxPattern++)
+        {
+            btPatterns |= (1 << (optRngPattern.optGetPattern(idxPattern) - minPattern));
```

Comment:
I am not sure I understand why do you need to calculate `btPatterns` here - I though the idea is just to emit a SWITCH and let the jit handle it, you only need to make sure switch has less than 64 cases (or some other profitability checks)

Reply:
Per offline discussion, skipped matching blocks with side effects to allow the change of block order. And, removed bitmap.